File: src/workers/continuum-core/bindings/modules/tool_parsing.ts
/**
 * RustCoreIPC Tool Parsing Module - Tool call parsing, correction, codec
 *
 * Replaces 784 lines of TypeScript ToolFormatAdapter hierarchy with Rust IPC.
 * Five format adapters (Anthropic XML, function-style, bare JSON, markdown backtick,
 * old-style XML) + parameter correction + tool name codec.
 */

import type { RustCoreIPCClientBase } from './base';
import type {
	ToolParseResult,
	ParsedToolCall,
	CorrectedToolCall,
} from '../../../../shared/generated';

export interface ToolParsingMixin {
	/**
	 * Parse tool calls from AI response text using all 5 format adapters.
	 * Returns parsed+corrected tool calls and cleaned text with tool blocks stripped.
	 * Sub-microsecond in Rust.
	 */
	toolParsingParse(responseText: string): Promise<ToolParseResult>;

	/**
	 * Correct a single tool call: name mapping + parameter mapping + content cleaning.
	 */
	toolParsingCorrect(toolName: string, parameters: Record<string, string>): Promise<CorrectedToolCall>;

	/**
	 * Register tool names for codec reverse lookup.
	 * Call once at startup with all known tool names.
	 */
	toolParsingRegisterTools(tools: string[]): Promise<{ registered: number; total: number }>;

	/**
	 * Decode a model-produced tool name variant back to the original.
	 * Handles: code_write, code__write, $FUNCTIONS.code_write, code-write, etc.
	 */
	toolParsingDecodeName(name: string): Promise<{ decoded: string; changed: boolean }>;

	/**
	 * Encode a tool name for API transmission: slashes -> underscores.
	 */
	toolParsingEncodeName(name: string): Promise<{ encoded: string }>;
}

export function ToolParsingMixin<T extends new (...args: any[]) => RustCoreIPCClientBase>(Base: T) {
	return class extends Base implements ToolParsingMixin {
		async toolParsingParse(responseText: string): Promise<ToolParseResult> {
			const response = await this.request({
				command: 'tool-parsing/parse',
				response_text: responseText,
			});

			if (!response.success) {
				throw new Error(response.error || 'Failed to parse tool calls');
			}

			return response.result as ToolParseResult;
		}

		async toolParsingCorrect(toolName: string, parameters: Record<string, string>): Promise<CorrectedToolCall> {
			const response = await this.request({
				command: 'tool-parsing/correct',
				tool_name: toolName,
				parameters,
			});

			if (!response.success) {
				throw new Error(response.error || 'Failed to correct tool call');
			}

			return response.result as CorrectedToolCall;
		}

		async toolParsingRegisterTools(tools: string[]): Promise<{ registered: number; total: number }> {
			const response = await this.request({
				command: 'tool-parsing/register-tools',
				tools,
			});

			if (!response.success) {
				throw new Error(response.error || 'Failed to register tools');
			}

			return response.result as { registered: number; total: number };
		}

		async toolParsingDecodeName(name: string): Promise<{ decoded: string; changed: boolean }> {
			const response = await this.request({
				command: 'tool-parsing/decode-name',
				name,
			});

			if (!response.success) {
				throw new Error(response.error || 'Failed to decode tool name');
			}

			return response.result as { decoded: string; changed: boolean };
		}

		async toolParsingEncodeName(name: string): Promise<{ encoded: string }> {
			const response = await this.request({
				command: 'tool-parsing/encode-name',
				name,
			});

			if (!response.success) {
				throw new Error(response.error || 'Failed to encode tool name');
			}

			return response.result as { encoded: string };
		}
	};
}

File: src/workers/continuum-core/src/modules/tool_parsing.rs
//! ToolParsingModule — stateless tool call parsing + correction IPC.
//!
//! Commands:
//! - `tool-parsing/parse`: Parse response text -> tool calls + cleaned text
//! - `tool-parsing/correct`: Correct a single tool call (name + params)
//! - `tool-parsing/register-tools`: Register tool names for codec
//! - `tool-parsing/decode-name`: Decode a model-produced tool name
//! - `tool-parsing/encode-name`: Encode a tool name for API transmission

use crate::runtime::{ServiceModule, ModuleConfig, ModulePriority, CommandResult, ModuleContext};
use crate::tool_parsing::{self, ToolNameCodec};
use crate::utils::params::Params;
use async_trait::async_trait;
use serde_json::Value;
use std::any::Any;
use std::sync::Arc;
use std::collections::HashMap;

pub struct ToolParsingModule {
    codec: Arc<ToolNameCodec>,
}

impl ToolParsingModule {
    pub fn new() -> Self {
        Self {
            codec: Arc::new(ToolNameCodec::new()),
        }
    }
}

#[async_trait]
impl ServiceModule for ToolParsingModule {
    fn config(&self) -> ModuleConfig {
        ModuleConfig {
            name: "tool-parsing",
            priority: ModulePriority::Normal,
            command_prefixes: &["tool-parsing/"],
            event_subscriptions: &[],
            needs_dedicated_thread: false,
            max_concurrency: 0,
            tick_interval: None,
        }
    }

    async fn initialize(&self, _ctx: &ModuleContext) -> Result<(), String> {
        Ok(())
    }

    async fn handle_command(
        &self,
        command: &str,
        params: Value,
    ) -> Result<CommandResult, String> {
        let p = Params::new(&params);

        match command {
            "tool-parsing/parse" => {
                let response_text = p.str("response_text")?;
                let result = tool_parsing::parse_and_correct(response_text);
                CommandResult::json(&result)
            }

            "tool-parsing/correct" => {
                let tool_name = p.str("tool_name")?;
                let parameters: HashMap<String, String> = match params.get("parameters") {
                    Some(Value::Object(map)) => {
                        map.iter().map(|(k, v)| {
                            (k.clone(), match v {
                                Value::String(s) => s.clone(),
                                _ => v.to_string(),
                            })
                        }).collect()
                    }
                    _ => HashMap::new(),
                };
                let corrected = tool_parsing::correction::correct_tool_call(tool_name, &parameters);
                CommandResult::json(&corrected)
            }

            "tool-parsing/register-tools" => {
                let tools: Vec<String> = match params.get("tools") {
                    Some(Value::Array(arr)) => {
                        arr.iter().filter_map(|v| v.as_str().map(String::from)).collect()
                    }
                    _ => return Err("Missing 'tools' array".to_string()),
                };
                let count = tools.len();
                self.codec.register_all(&tools);
                Ok(CommandResult::Json(serde_json::json!({
                    "registered": count,
                    "total": self.codec.count(),
                })))
            }

            "tool-parsing/decode-name" => {
                let raw = p.str("name")?;
                let decoded = self.codec.decode(raw);
                Ok(CommandResult::Json(serde_json::json!({
                    "decoded": decoded,
                    "changed": decoded != raw,
                })))
            }

            "tool-parsing/encode-name" => {
                let name = p.str("name")?;
                let encoded = self.codec.encode(name);
                Ok(CommandResult::Json(serde_json::json!({
                    "encoded": encoded,
                })))
            }

            _ => Err(format!("Unknown command: {command}")),
        }
    }

    fn as_any(&self) -> &dyn Any { self }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn test_parse_command() {
        let module = ToolParsingModule::new();
        let params = serde_json::json!({
            "response_text": "<tool_use><tool_name>code/search</tool_name><parameters><query>test</query></parameters></tool_use>"
        });
        let result = module.handle_command("tool-parsing/parse", params).await;
        assert!(result.is_ok());
        if let Ok(CommandResult::Json(json)) = result {
            let calls = json["tool_calls"].as_array().unwrap();
            assert_eq!(calls.len(), 1);
            assert_eq!(calls[0]["tool_name"], "code/search");
            // query -> pattern (correction)
            assert!(calls[0]["parameters"]["pattern"].is_string());
        }
    }

    #[tokio::test]
    async fn test_correct_command() {
        let module = ToolParsingModule::new();
        let params = serde_json::json!({
            "tool_name": "workspace/tree",
            "parameters": { "directory": "./src" }
        });
        let result = module.handle_command("tool-parsing/correct", params).await;
        assert!(result.is_ok());
        if let Ok(CommandResult::Json(json)) = result {
            assert_eq!(json["tool_name"], "code/tree");
            assert_eq!(json["name_changed"], true);
            assert_eq!(json["parameters"]["path"], "./src");
        }
    }

    #[tokio::test]
    async fn test_register_and_decode() {
        let module = ToolParsingModule::new();

        // Register tools
        let reg_params = serde_json::json!({
            "tools": ["code/write", "code/read", "collaboration/chat/send"]
        });
        let reg_result = module.handle_command("tool-parsing/register-tools", reg_params).await;
        assert!(reg_result.is_ok());
        if let Ok(CommandResult::Json(json)) = reg_result {
            assert_eq!(json["registered"], 3);
            assert_eq!(json["total"], 3);
        }

        // Decode encoded name
        let dec_params = serde_json::json!({ "name": "code_write" });
        let dec_result = module.handle_command("tool-parsing/decode-name", dec_params).await;
        assert!(dec_result.is_ok());
        if let Ok(CommandResult::Json(json)) = dec_result {
            assert_eq!(json["decoded"], "code/write");
            assert_eq!(json["changed"], true);
        }

        // Decode with prefix
        let prefix_params = serde_json::json!({ "name": "$FUNCTIONS.code_write" });
        let prefix_result = module.handle_command("tool-parsing/decode-name", prefix_params).await;
        assert!(prefix_result.is_ok());
        if let Ok(CommandResult::Json(json)) = prefix_result {
            assert_eq!(json["decoded"], "code/write");
        }
    }

    #[tokio::test]
    async fn test_encode_command() {
        let module = ToolParsingModule::new();
        let params = serde_json::json!({ "name": "collaboration/chat/send" });
        let result = module.handle_command("tool-parsing/encode-name", params).await;
        assert!(result.is_ok());
        if let Ok(CommandResult::Json(json)) = result {
            assert_eq!(json["encoded"], "collaboration_chat_send");
        }
    }

    #[tokio::test]
    async fn test_unknown_command() {
        let module = ToolParsingModule::new();
        let result = module.handle_command("tool-parsing/nope", Value::Null).await;
        assert!(result.is_err());
    }
}
